package/src/resolve.rs

```diff
@@ -0,0 +1,329 @@
+use std::collections::HashMap;
+
+use nickel_lang_core::{cache::normalize_path, identifier::Ident, package::PackageMap};
+
+use crate::{
+    config::Config,
+    error::{Error, IoResultExt as _},
+    index::{self, PackageIndex, Shared},
+    lock::{LockFile, LockPrecisePkg},
+    snapshot::Snapshot,
+    version::{SemVer, VersionReq},
+    Dependency, IndexDependency, ManifestFile, PrecisePkg,
+};
+
+/// Stores the result of resolving version constraints to exact versions.
+#[derive(Debug)]
+pub struct Resolution {
+    pub config: Config,
+    /// The snapshot (of path and git packages) that was used to construct
+    /// this resolution. Note that path and git packages are not "resolved";
+    /// they have fixed versions. The snapshot is only used to collect index
+    /// dependencies of git and path packages.
+    pub snapshot: Snapshot,
+    pub index: PackageIndex<Shared>,
+    /// All the index packages in the dependency tree.
+    ///
+    /// Each package id can resolve to multiple versions, but those versions should all fall
+    /// into disjoint semantic-version buckets.
```

Comment:
Is it true even when using precise version bounds? What happens if two dependencies define two different but compatible (in the same bucket) exact version bounds of a third dependency? Or maybe exact version bounds are just not recorded here in `index_packages`?

Reply:
In this PR, two different but compatible version bounds are allowed (because we basically don't check anything). The next PR will disallow it -- it will ensure when populating `index_packages` that there are no conflicts.

Reply:
So if we have two different dependencies A and B that defined compatible but different exact version bounds on the C dependencies, say "1.1.0" and "1.2.1", we can't just use A and B together? Is that expected/part of the cargo semantics as well?